server/src/util/Storage.ts

```diff
@@ -0,0 +1,136 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Hvy Industries. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *  "HVY", "HVY Industries" and "Hvy Industries" are trading names of JCKD (UK) Ltd
+ *--------------------------------------------------------------------------------------------*/
+
+const fs = require("fs");
+const zlib = require('zlib');
+const util = require('util');
+
+let isProcessing = false;
+let waitProcessing = null;
+let waitInterval = 5000;
+let errHandler = null;
+let pendingCb = [];
+
+/**
+ * Requests to flush the specified tree structure
+ */
+export function onError(cb: (err: Error) => void): void {
+    errHandler = cb;
+};
+
+/**
+ * Raise an error
+ */
+function raiseError(err: Error): void {
+    if (errHandler) {
+        errHandler(err);
+    } else {
+        console.error(err.stack ? err.stack : err);
+    }
+};
+
+/**
+ * Reads the specified filename
+ */
+export function read(filename: string, cb: (err?: Error, tree?:any) => void): void {
+    if (isProcessing) {
+        // bad things may happen, so delay it by 1 sec
+        setTimeout(() => {
+            read(filename, cb);
+        }, 1000);
```

Comment:
Yes, you are right. I was using 1 sec in order to avoid one edge case, but it's a bad way to handle it.

I think I should use a readFileSync instead, because event if I start before the flag is set, I could potentially read, and write at the same time as actions are async.

Anyway it's a false problem here, as the read function is used at start when the tree is not yet loaded, so there is no way to have a collision in actual system.